Better TextSnippet support for moving cursor & deleting text

lost-werewolf opened the pull request:
Hello! I really like this mod, it's a great overhaul to vanilla's chat system. I was mildly annoyed by TextSnippets not being deleted fully when pressing backspace, the cursor not moving fully across snippets, and ctrl-clicking items into chat not moving the cursor.
This PR aims to remedy those issues:

- Moving the cursor left/right will move the cursor the full length of a snippet if it's set to DeleteWhole
- Pressing backspace will delete entire snippets (respects DeleteWhole)
- Inserting snippets through ChatManager.AddChatText (such as clicking an item into chat) will advance the cursor the length of the snippet

I apologize if my code style/quality is not up to par, if there's anything I didn't do right or could have done better, feel free to let me know! I tried to follow your formatting to the best of my ability, but I may have fallen flat in some places.

lost-werewolf committed:
Better TextSnippet support

Support TextSnippets when moving the caret left & right
Support TextSnippets when pressing backspace
Detour ChatManager.AddChatText to move the caret when adding a snippet into chat (such as ctrl-clicking an item into chat)

lost-werewolf committed:
Formatting & Cleanup

emyhrberg commented:
@lost-werewolf  This looks good, I'll test in a few hours and merge if all looks well

lost-werewolf committed:
Re-Implement ChatManager.AddChatText

ChatManager.AddChatText will now add the text at the cursor position rather than adding to the end of the string

lost-werewolf commented:
Just fixed something I totally missed in the first pass. I hope this is acceptable! Needed to entirely override ChatManager.AddChatText so it inserts at the cursor position.

emyhrberg commented:
> Just fixed something I totally missed in the first pass. I hope this is acceptable! Needed to entirely override ChatManager.AddChatText so it inserts at the cursor position.

Alright, I'll have a look. Care to continue on Discord for future collaboration? My name is erkyy on discord.

emyhrberg commented:
@lost-werewolf Hi, it seems the files `HandleChatSystem` an d `HandleChatDetourSystem` were placed in the wrong folder, it would be good if they were placed in the right folder.

Additionally, if you could squash your commits into one that would also be an improvement to easily see the changes.

I'm working to fix it right now.

emyhrberg commented:
@lost-werewolf Just pushed your code

Here's how it looks in-game now. I assume this is what your PR fixed but feel free to clarify if there is something missing.
https://github.com/user-attachments/assets/fd1fae99-5ba6-4ae2-90e4-d282231b9f97

I'm making a new commit here with your changes: https://github.com/emyhrberg/ChatPlus/commit/dc98f28801235848118361dd1f040c35a70a084f

Note that there was nothing wrong with your code, but your PR made it so when I merged it, it created a duplicate file of `HandleChatSystem` since it was not placed in the right folder. You can check out the commit (see above) to see how it should have done for next time.

lost-werewolf commented:
Sorry, was busy with IRL appointments the past two weeks! Didn't see that it caused that. I am a bit clumsy, so I likely missed that. Sorry!

Regarding the attachment, yes, that is what my PR fixed. Snippets are a bit confusing sometimes so I learned how to properly handle them to avoid frustration on the user's end. Especially since tModLoader changes how item tags are sent in chat, which results in a very long string that without proper support, is annoying to navigate through.

I think the only missing thing would be to render a TextSnippet's original text instead of the formatted text/custom draw when the cursor intersects it somehow (either by manually clicking inside of it or if the snippet isn't DeleteWhole). This is something I am unsure of how to approach, so that's why I didn't attempt it.
